File: solutions/csharp/perfect-numbers/1/PerfectNumbers.cs
public enum Classification
{
    Perfect,
    Abundant,
    Deficient
}

public static class PerfectNumbers
{
    public static Classification Classify(int number)
    {
        if (number < 1)
        {
            throw new ArgumentOutOfRangeException(nameof(number));
        }

        return AliquotSum(number).CompareTo(number) switch
        {
            0 => Classification.Perfect,
            > 0 => Classification.Abundant,
            _ => Classification.Deficient
        };
    }

    private static int AliquotSum(int number)
    {
        if (number == 1) return 0;

        return Enumerable
            .Range(2, (int)Math.Sqrt(number) - 1)
            .Where(IsMultipleOf(number))
            .SelectMany(FactorPair(number))
            .Sum() + 1;
    }

    private static Func<int, bool> IsMultipleOf(int number) => factor => number % factor == 0;

    private static Func<int, IEnumerable<int>> FactorPair(int number) =>
        factor =>
        {
            var pair = number / factor;

            return pair == factor ? [factor] : [factor, pair];
        };
}

File: solutions/scala/perfect-numbers/1/src/main/scala/PerfectNumbers.scala
enum NumberType {
  case Perfect, Abundant, Deficient
}

object PerfectNumbers {
  private val InvalidNumber = "Classification is only possible for natural numbers."

  def classify(number: Int): Either[String, NumberType] =
    if number < 1 then Left(InvalidNumber)
    else Right(number.compareTo(aliquotSum(number)) match {
      case 0 => NumberType.Perfect
      case 1 => NumberType.Deficient
      case -1 => NumberType.Abundant
    })

  private def aliquotSum(number: Int): Int =
    if number == 1 then 0
    else
      val root = math.sqrt(number).toInt
      val pairedFactors = (2 to root).filter(number % _ == 0).flatMap(factor => pairedFactor(number, factor))

      (1 +: pairedFactors).sum

  private def pairedFactor(number: Int, factor: Int): Seq[Int] =
    val pair = number / factor

    if factor == pair then Seq(factor)
    else Seq(factor, pair)
}
